File: src/MyAgent.Orchestrator/Agent/HumanLoop.cs
using System.Text.Json.Nodes;
using MyAgent.Orchestrator.Mcp;

namespace MyAgent.Orchestrator.Agent;

public static class HumanLoop
{
    private const string Bold = "\x1b[1m";
    private const string Yellow = "\x1b[93m";
    private const string Reset = "\x1b[0m";

    public static string Ask(string question)
    {
        Console.Write($"\n{Bold}{Yellow}❓  Agent asks: {Reset}{question}\n{Bold}👤  Your answer: {Reset}");
        return Console.ReadLine() ?? "";
    }

    public static McpToolDefinition GetToolDefinition() => new()
    {
        Name = "ask_human",
        Description = "Ask the human engineer a question when you need clarification, approval, or are stuck. Use this tool freely — the human is always available.",
        InputSchema = new JsonObject
        {
            ["type"] = "object",
            ["properties"] = new JsonObject
            {
                ["question"] = new JsonObject
                {
                    ["type"] = "string",
                    ["description"] = "The question or prompt to present to the human."
                }
            },
            ["required"] = new JsonArray("question")
        }
    };
}

File: agent/human_loop.py
"""Human-in-the-loop prompt helper."""

# ANSI colour codes
_CYAN = "\033[96m"
_YELLOW = "\033[93m"
_RESET = "\033[0m"
_BOLD = "\033[1m"


def ask_human(question: str) -> str:
    """Print a formatted question to the terminal and return the user's answer.

    Args:
        question: The question the agent wants to ask the human operator.

    Returns:
        The stripped text entered by the user.
    """
    print(f"\n{_BOLD}{_CYAN}{'=' * 60}{_RESET}")
    print(f"{_BOLD}{_CYAN}🤖  AGENT NEEDS YOUR INPUT{_RESET}")
    print(f"{_BOLD}{_CYAN}{'=' * 60}{_RESET}")
    print(f"\n{_YELLOW}{question}{_RESET}\n")
    answer = input(f"{_BOLD}👉  Your answer: {_RESET}").strip()
    print(f"{_CYAN}{'=' * 60}{_RESET}\n")
    return answer
